File: src/main/java/com/rtc/app/auth/dto/internal/CustomUserDetails.java
package com.rtc.app.auth.dto.internal;

import com.rtc.app.auth.entity.User;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.userdetails.UserDetails;

import java.util.Collection;
import java.util.List;

public class CustomUserDetails implements UserDetails {

    private final Long id;
    private final String email;
    private final String password;
    private final Collection<? extends GrantedAuthority> authorities;

    public CustomUserDetails(Long id, String email, String password, Collection<? extends GrantedAuthority> authorities) {
        this.id = id;
        this.email = email;
        this.password = password;
        this.authorities = authorities;
    }

    public static CustomUserDetails from(User user) {
        List<GrantedAuthority> authorities = List.of(() -> user.getType().getSecurityLevel());
        return new CustomUserDetails(user.getId(), user.getEmail(), user.getPassword(), authorities);
    }

    public Long getId() {
        return id;
    }

    public String getEmail() {
        return email;
    }

    @Override
    public Collection<? extends GrantedAuthority> getAuthorities() {
        return authorities;
    }

    @Override
    public String getPassword() {
        return password;
    }

    @Override
    public String getUsername() {
        return email;
    }
}

File: src/main/kotlin/com/rtc/app/auth/dto/internal/CustomUserDetails.kt
package com.rtc.app.auth.dto.internal

import com.rtc.app.auth.entity.User
import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.core.userdetails.UserDetails

class CustomUserDetails(
    val id: Long,
    val email: String,
    private val password: String?,
    private val authorities: MutableCollection<out GrantedAuthority?>?
) : UserDetails {

    companion object {

        @JvmStatic
        fun from(user: User): CustomUserDetails {
            val authorities = listOf(SimpleGrantedAuthority(user.type.securityLevel))

            return CustomUserDetails(
                id = user.id!!,
                email = user.email,
                password = user.password,
                authorities = authorities.toMutableList()
            )
        }
    }

    override fun getAuthorities(): Collection<GrantedAuthority> = authorities as Collection<GrantedAuthority>

    override fun getPassword(): String? = password

    override fun getUsername(): String = email
}